### PR TITLE
P2 draft

### PR DESCRIPTION
**Pull request by shenghaozou**
Please let us know if you have any questions.

### CODE REVIEW

File: fall18/p2/README.md [hunk 1]
```diff
@@ -1,3 +1,149 @@
 # Project 2
 
-More details coming soon.
+In this project, you'll learn about creating and calling functions, arguments, return values and math module. To start, download these files:
```
Comment (tylerharter) [inline fall18/p2/README.md]: Let's add a note saying to right click the links below, not the links above that only show a preview

File: fall18/p2/README.md [hunk 2]
```diff
@@ -1,3 +1,149 @@
...
+|Illinois|57913.55 |12419293|12830632|12882189|
+
+</center>
+
+1. What is the value of the maximum land area among any three given states?
+2. What is the least value of population density among any three given states?
+3. For a given state, what’s the predicted population in the given year? (The function will take in values of initial year, final year, growth rate and state name)
+4. Between any two given years, what is the population growth rate of particular state?
+
+### Provided Function in Module project.py
```
Comment (tylerharter) [inline fall18/p2/README.md]: typo: Functions

File: fall18/p2/README.md [hunk 3]
```diff
@@ -1,3 +1,149 @@
...
+  Problem 11: function predictPopulation not found
+  Problem 12: function predictPopulation not found
+  Problem 13: function predictPopulation not found
+  Problem 14: function predictPopulation not found
+  Problem 15: function predictPopulation not found
+  Problem 16: function calcGrowthRate not found
+  Problem 17: function calcGrowthRate not found
+  Problem 18: function calcGrowthRate not found
+  Problem 19: function calcGrowthRate not found
+  Problem 20: function calcGrowthRate not found
```
Comment (tylerharter) [inline fall18/p2/README.md]: When I run it, I also see a total score at the bottom.

File: fall18/p2/project.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
+import csv
+
+
+def getArea(stateName):
+    """ Function that takes in state name and returns the area of the state"""
+    with open('area.csv') as csvDataFile:
+        csvReader = csv.reader(csvDataFile)
+        for row in csvReader:
+            if row[0] == stateName:
```
Comment (tylerharter) [inline fall18/p2/project.py]: No need to change this, but DictReader is a cleaner way to do this without numerical indexes.

File: fall18/p2/README.md [hunk 5]
```diff
@@ -1,3 +1,149 @@
...
+
+| State | Area (mi²) | Population (2000) | Population (2010) | population (2014) |
+|------|------|------|------|------|
+|Wisconsin|65496.38 |5363675|5686986|5759432|
+|Minnesota|86935.83 |44919479|5303925|5457125|
+|Illinois|57913.55 |12419293|12830632|12882189|
+
+</center>
+
+1. What is the value of the maximum land area among any three given states?
```
Comment (tylerharter) [inline fall18/p2/README.md]: There should be a sentence before this, saying what these questions are

File: fall18/p2/README.md [hunk 6]
```diff
@@ -1,3 +1,149 @@
...
+>>> 86935.83
+```
+
+**getMinimumPopulationDensity(stateName1, stateName2, stateName3, year)**
+
+Population density is measured as population per unit area. 
+
+The formula to calculate population density is:
+
+**Population Density = Population / Land Area (mi²)**
```
Comment (tylerharter) [inline fall18/p2/README.md]: It's weird that this has the same styling as the four functions.  Maybe only the four functions should be bold?  Or, separate them with some extra text, like this:

Function 1: getMaximumLand(...)

File: fall18/p2/README.md [hunk 7]
```diff
@@ -1,3 +1,149 @@
...
+Example:
+
+```
+>>> predictPopulation("Wisconsin", 0.5, 2000, 2010)
+>>> 796039951.1495125
+```
+
+**calcGrowthRate(stateName, yearA, yearB)**
+
+Use the same formula as above to calculate the growth rate. To do this, you can use the log fucntion present in Python [math library](https://docs.python.org/3/library/math.html).
```
Comment (tylerharter) [inline fall18/p2/README.md]: misspelled function

File: fall18/p2/area.csv [hunk 8]
```diff
@@ -0,0 +1,51 @@
+Alabama,52420.07
```
Comment (tylerharter) [inline fall18/p2/area.csv]: Nice, glad to see data for all the states!

File: fall18/p2/README.md (hunk 7 above)
Comment (tylerharter) [inline fall18/p2/README.md]: Maybe we should give them a bit more of a hint here.  For example, remind them that log(e**(rt)) == rt, and tell them they need to rearrange the above formula to isolate r.  Sadly, I don't think we can assume the freshmen in our classes will all be leaving high school comfortable working with logs.

File: fall18/p2/main.py [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+# Insert calcGrowthRate(stateName, yearA, yearB) function here
+# Step 1: Using the statename retrieve the population for yearA
+# Step 1: Using the statename retrieve the population for yearB
+# Step 2: Calculate time between yearA and yearB
+# Step 3: Use the math.log() function in the given formula to calculate growth rate
+# Step 4: Return the final population
+
+
+if __name__ == "__main__":
```
Comment (tylerharter) [inline fall18/p2/main.py]: I think it's cleaner to do something like this:

def main():
   ...

if __name__ == "__main__":
    main()

File: fall18/p2/main.py [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    # You don't need to change anything below this line
+    state1Name = "Wisconsin"
+    state2Name = "Iowa"
+    state3Name = "Minnesota"
+
+    # Solving Q1: Finding max Area
+    # Here we call the function and see if it gives the correct output
+    # Play around by changing the state names and your program should still work.
+    maxArea = getMaximumLand(state1Name, state2Name, state3Name)
+    print ("The maximum land area among the three states %s, %s and %s : " % (state1Name, state2Name, state3Name))
```
Comment (tylerharter) [inline fall18/p2/main.py]: They don't know about string formatting yet.  Let's rephrase this to use "+" instead so they can read it more easily.

File: fall18/p2/main.py [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+    print ("The minimum population density among the three states %s, %s and %s : " % (state1Name, state2Name, state3Name))
+    print (minPopulationDensity)
+
+    # Solving Q3: Finding final population
+    # Play around by changing the state names and the year and your program should still work.
+    finalPopulation = predictPopulation(state1Name, 0.5, 2000,2010)
+    print ("The new population for %s : " % state1Name)
+    print (finalPopulation)
+
+    # Solving Q4: Finding minimum Population Density
```
Comment (tylerharter) [inline fall18/p2/main.py]: Oops, this is the same as Q2

File: fall18/p2/main.py [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+# Insert calcGrowthRate(stateName, yearA, yearB) function here
+# Step 1: Using the statename retrieve the population for yearA
+# Step 1: Using the statename retrieve the population for yearB
+# Step 2: Calculate time between yearA and yearB
+# Step 3: Use the math.log() function in the given formula to calculate growth rate
+# Step 4: Return the final population
+
+
+if __name__ == "__main__":
+    # Example state names. Please play around by changing the state names
```
Comment (tylerharter) [inline fall18/p2/main.py]: it would be nice if they had to write a couple things in main that gets tested (with check_output, as in P1).  These would just call our functions, and would give them some points even before they create their own functions

File: fall18/p2/test.py [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+import subprocess, json, re, sys
+from inspect import getmembers, isfunction
+import main
+
+PASS = 'PASS'
+PROBLEMS = 20
+EPSILON = 0.0001
+
+TESTS = [
+    ("getMaximumLand", ["Wisconsin", "Iowa", "Minnesota"], 86935.83), #1-1
```
Comment (tylerharter) [inline fall18/p2/test.py]: Hmmm, there are really only 4 distinct tests here.  I have trouble imaging how a student could get a grade other than 100%, 75%, 50%, etc...

Comment (shenghaozou) [reply fall18/p2/test.py]: Yes... If you think 4 tests are better I can try that.

File: fall18/p2/main.py (hunk 12 above)
Comment (shenghaozou) [reply fall18/p2/main.py]: I am not sure if I have enough to update because we need to release at Wed but this requires to change test, main and markdown file. Let me deal with other problem first.

File: fall18/p2/main.py (hunk 10 above)
Comment (shenghaozou) [reply fall18/p2/main.py]: Good suggestion. :)

File: fall18/p2/main.py (hunk 9 above)
Comment (shenghaozou) [reply fall18/p2/main.py]: I'll fix soon.

File: dockerGrader/testServer.py [hunk 14]
```diff
@@ -3,7 +3,7 @@
```
[the remark this answers is not in the document]
Comment (shenghaozou) [reply dockerGrader/testServer.py]: Oh yes... I'll git rm them.

File: fall18/p2/README.md [hunk 15]
```diff
@@ -1,3 +1,149 @@
...
+
+* [main.py](https://raw.githubusercontent.com/shenghaozou/taDebug/master/main.py)
+* [project.py](https://raw.githubusercontent.com/shenghaozou/taDebug/master/project.py)
+* [test.py](https://raw.githubusercontent.com/shenghaozou/taDebug/master/test.py)
+* [area.csv](https://raw.githubusercontent.com/shenghaozou/taDebug/master/area.csv)
+* [population.csv](https://raw.githubusercontent.com/shenghaozou/taDebug/master/population.csv)
+
+You will change main.py and hand it in. You should not change test.py, and you should not hand it in; it's only purpose is to tell you your grade in advance.
+
+After you've downloaded both files to the same directory, open your terminal, navigate to that directory, and run the following:
```
Comment (adalbertgerald) [inline fall18/p2/README.md]: Change "both files" to "all the above 5 files" since there are more than two files.

File: fall18/p2/README.md [hunk 16]
```diff
@@ -1,3 +1,149 @@
...
+2. What is the least value of population density among any three given states?
+3. For a given state, what’s the predicted population in the given year? (The function will take in values of initial year, final year, growth rate and state name)
+4. Between any two given years, what is the population growth rate of particular state?
+
+### Provided Function in Module project.py
+We don’t require you to know how to read from data file so we have provided the following functions. Please make sure the data file area.csv and population.csv are in the same folder along with main.py.
+
+**getArea(stateName)** 
+
+getArea function takes the name of one state as an argument and returns the size of this state area.
```
Comment (adalbertgerald) [inline fall18/p2/README.md]: It would be better to say "returns the area of this state."

File: fall18/p2/main.py [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+
+# Insert getMaximumLand(stateName1, stateName2, stateName3) function here
+# Step1: Retrieve area values for all 3 states using getArea function and assign it to variables.
+# Step2: Use max() function to calculate the max value and store it as a variable
+# Step3: Return the max value that you calculated.
+
+
+# Insert getMinimumPopulationDensity(stateName1, stateName2, stateName3, year) function here
+# Step1: Retrieve area and population values for all 3 states using getArea and the getPopulation function and assign it to variables.
+# Step2: Use min() function to calculate the max value and store it as a variable
```
Comment (adalbertgerald) [inline fall18/p2/main.py]: Typo: "max value" -> "min value"

File: fall18/p2/main.py [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+# Step3: Return the min value that you calculated.
+
+
+# Insert predictPopulation(stateName, growthRate, yearA, yearB): function here
```
Comment (adalbertgerald) [inline fall18/p2/main.py]: Remove the colon (:) after function to be consistent with other functions.

File: fall18/p2/main.py [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+# Step 1: Using the statename retrieve the population for yearA
+# Step 2: Calculate time between yearA and yearB
+# Step 3: Use the math.exp() function in the given formula to calculate final population
+# Step 4: Return the predicted population
```
Comment (adalbertgerald) [inline fall18/p2/main.py]: Better to say "return the predicted final population".

File: fall18/p2/main.py [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+
+
+# Insert calcGrowthRate(stateName, yearA, yearB) function here
+# Step 1: Using the statename retrieve the population for yearA
+# Step 1: Using the statename retrieve the population for yearB
+# Step 2: Calculate time between yearA and yearB
+# Step 3: Use the math.log() function in the given formula to calculate growth rate
+# Step 4: Return the final population
```
Comment (adalbertgerald) [inline fall18/p2/main.py]: Change "final population" to "growth rate".

File: fall18/p2/main.py [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+
+    # Solving Q2: Finding minimum Population Density
+    # Play around by changing the state names and the year and your program should still work.
+    minPopulationDensity = getMinimumPopulationDensity(state1Name, state2Name, state3Name, 2000)
+    print ("The minimum population density among the three states %s, %s and %s : " % (state1Name, state2Name, state3Name))
+    print (minPopulationDensity)
+
+    # Solving Q3: Finding final population
+    # Play around by changing the state names and the year and your program should still work.
+    finalPopulation = predictPopulation(state1Name, 0.5, 2000,2010)
```
Comment (adalbertgerald) [inline fall18/p2/main.py]: Suggestion: add a space after the comma between 2000 and 2010.

File: fall18/p2/project.py [hunk 22]
```diff
@@ -0,0 +1,25 @@
...
+    """ Function that takes in state name and returns the area of the state"""
+    with open('area.csv') as csvDataFile:
+        csvReader = csv.reader(csvDataFile)
+        for row in csvReader:
+            if row[0] == stateName:
+                return float(row[1])
+
+
+def getPopulation(stateName,year):
+    """ Function that takes in state name and year to returns the population for that year"""
```
Comment (adalbertgerald) [inline fall18/p2/project.py]: Typo: "to returns" -> "and returns"

File: fall18/p2/project.py [hunk 23]
```diff
@@ -0,0 +1,25 @@
...
+    with open('population.csv') as csvDataFile:
+        csvReader = csv.reader(csvDataFile)
+        row0= next(csvReader)
```
Comment (adalbertgerald) [inline fall18/p2/project.py]: Style: Add space between row0 and =

File: fall18/p2/main.py (hunk 9 above)
Comment (shrunambiar) [reply fall18/p2/main.py]: Yes. Will do this.

File: fall18/p2/main.py (hunk 10 above)
Comment (shrunambiar) [reply fall18/p2/main.py]: Taken care of.

File: fall18/p2/README.md [hunk 24]
```diff
@@ -1,3 +1,149 @@
 # Project 2
 
-More details coming soon.
+In this project, you'll learn about creating and calling functions, arguments, return values and math module. To start, download these files:
+
+* [main.py](https://raw.githubusercontent.com/shenghaozou/taDebug/master/main.py)
```
[the remark this answers is not in the document]
Comment (shenghaozou) [reply fall18/p2/README.md]: Yeah but not now. It should appear in another pull request.

File: fall18/p2/README.md (hunk 6 above)
Comment (adalbertgerald) [reply fall18/p2/README.md]: I agree. Please remove the bold from the formula for Population Density and add a prefix (e.g., Function 1) before the functions.

File: fall18/p2/README.md (hunk 7 above)
Comment (adalbertgerald) [reply fall18/p2/README.md]: Please fix this typo.

Comment (adalbertgerald) [reply fall18/p2/README.md]: Yes, I too think that we should have this additional information here. @shenghaozou / @shrunambiar can you please add this?

File: fall18/p2/main.py [hunk 25]
```diff
@@ -0,0 +1,62 @@
...
+    minDensity = getMinimumPopulationDensity(state1Name, state2Name, state3Name, 2000)
+    print("Minimum population density among " + state1Name + "," + state2Name + "," + state3Name + " is " + minDensity)
+
+    # Solving Q3: Finding final population
+    # Play around by changing the state names and the year and your program should still work.
+    finalPopulation = predictPopulation(state1Name, 0.5, 2000, 2010)
+    print("The new population for " + state1Name + " is " + finalPopulation)
+
+
+    # Solving Q4: Finding minimum Population Density
```
Comment (adalbertgerald) [inline fall18/p2/main.py]: This should be "# Solving Q4: Calculating growth rate."

File: fall18/p2/main.py [hunk 26]
```diff
@@ -0,0 +1,62 @@
...
+    state2Name = "Iowa"
+    state3Name = "Minnesota"
+
+    # Solving Q1: Finding max Area
+    # Here we call the function and see if it gives the correct output
+    # Play around by changing the state names and your program should still work.
+    maxArea = getMaximumLand(state1Name, state2Name, state3Name)
+    print("Maximum land area among " + state1Name + "," + state2Name + "," + state3Name + " is " + maxArea)
+
+    # Solving Q2: Finding minimum Population Densit,
```
Comment (adalbertgerald) [inline fall18/p2/main.py]: Typo: "Densit," -> "Density"

File: fall18/p2/test.py (hunk 13 above)
Comment (tylerharter) [reply fall18/p2/test.py]: No, we need at least 10 real tests that exercise different things

Comment (tylerharter) [reply fall18/p2/test.py]: Right now, it feel like we only have 4 real tests

File: fall18/p2/test.py [hunk 27]
```diff
@@ -0,0 +1,174 @@
...
+    ["The population of Wisconsin in 2010 is ", 5686986.0], #2
+    ["Maximum land area among Wisconsin, Iowa, Minnesota is ", 86935.83], #3
+    ["Minimum population density among Wisconsin, Iowa, Minnesota is ", 52.002450206414075], #4
+    ["The predicted population for Wisconsin in year 2010 is (assume start yearA is 2000, growth rate is 0.5) ", 796039951.1495125], #5
+    ["The growth rate for Wisconsin between year 2000 and 2010 is ", 0.005853103209551789] #6
+]
+
+TESTS_FUNC_NUM = len(TESTS_FUNC)
+TESTS_PRINT_NUM = len(TESTS_PRINT)
+PROBLEMS = TESTS_FUNC_NUM + TESTS_PRINT_NUM
```
Comment (tylerharter) [inline fall18/p2/test.py]: It's confusing to me that this is 11 but there are 16 tests.  I'm struggling to map the failures to specific tests...

File: fall18/p2/README.md [hunk 28]
```diff
@@ -1,3 +1,147 @@
...
+  Problem 2: function getMinimumPopulationDensity not found
+  Problem 3: function predictPopulation not found
+  Problem 4: function predictPopulation not found
+  Problem 5: function calcGrowthRate not found
+  Problem 11: PASS
+  Problem 12: expected (5686986.0) but found (0)
+  Problem 13: expected (86935.83) but found (0)
+  Problem 14: expected (52.002450206414075) but found (0)
+  Problem 15: expected (796039951.1495125) but found (0)
+  Problem 16: expected (0.005853103209551789) but found (0)
```
Comment (tylerharter) [inline fall18/p2/README.md]: Why not add back some of your original tests to get a nice even 20?

File: fall18/p2/test.py [hunk 29]
```diff
@@ -0,0 +1,174 @@
...
+TESTS_FUNC = [
+    ("getMaximumLand", ["Arizona", "California", "Connecticut"], 163694.74), #1
+    ("getMinimumPopulationDensity", ["Wisconsin", "Iowa", "Minnesota", 2010], 54.13546968775862), #2
+    ("predictPopulation", ["Hawaii", 2000, 2010, 0.4], 66147678.89670547), #3
+    ("predictPopulation", ["Hawaii", 2000, 2010], 3293299.011605786), #4
+    ("calcGrowthRate", ["Hawaii", 2000, 2015], 0.010596535979501064) #5
+]
+
+TESTS_PRINT = [
+    ["The area of Wisconsin is ", 65496.38], #1
```
Comment (tylerharter) [inline fall18/p2/test.py]: Why is both this number 1 and an early test is also number 1?

File: fall18/p2/main.py [hunk 30]
```diff
@@ -0,0 +1,87 @@
...
+# Step 4: Return the growth rate.
+# Note: You shouldn't use print in this function.
+
+def main():
+    # Example state names. Please play around by changing the state names
+    # Fix the main code so it will print the correct value
+
+    # Problem 6
+    # Read Area Data
+    print('Problem 6')
```
Comment (tylerharter) [inline fall18/p2/main.py]: Would have been nice if these were numbered differently so that students start with the easy ones...

File: fall18/p2/test.py [hunk 31]
```diff
@@ -0,0 +1,174 @@
...
+        for line in output.split('\n'):
+            m = re.match(r'Problem (\d+)$', line)
+            if m:
+                problem_number = int(m.group(1))
+                problems[problem_number] = []
+            elif problem_number:
+                problems[problem_number].append(line)
+
+        for problemNum in range(6, 12):
+            testCase = TESTS_PRINT[problemNum - 6]
```
Comment (tylerharter) [inline fall18/p2/test.py]: Hmmm, this feels tricky.

File: fall18/p2/test.py (hunk 27 above)
Comment (shenghaozou) [reply fall18/p2/test.py]: I only have 11 tests here. What's  16 comes from?

File: fall18/p2/README.md (hunk 28 above)
Comment (shenghaozou) [reply fall18/p2/README.md]: Ok.

File: fall18/p2/test.py (hunk 29 above)
Comment (shenghaozou) [reply fall18/p2/test.py]: I'll update.
